:sparkles: Allocate dedicated host when a dedicated host doesn't exist

rvanderp3 opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Builds on #5548 to enable CAPA to allocate a dedicated host. This removes the requirement that a dedicated host exist prior to instance provisioning.  A core assumption of this PR is that there is a 1:1 mapping between an instance and the dedicated host on which it runs. Nothing in this implementation precludes scheduling multiple instances to a given dedicated host, but that is not the goal of this initial implementation.


**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits
- [x] includes emoji in title 
- [x] adds unit tests

**Release note**:
```
Adds support for provisioning a new instance on a newly allocated dedicated host. The dedicate host is associated only with the new instance.  When the instance is deprovisioned, the dedicated host will be released as well.
```

rvanderp3 commented on api/v1beta2/awsmachine_webhook.go:
```diff
@@ -469,14 +469,161 @@ func (r *AWSMachine) validateAdditionalSecurityGroups() field.ErrorList {
+			// First character must be a letter
+			if (char < 'a' || char > 'z') && (char < 'A' || char > 'Z') {
+				return false
+			}
+		} else {
+			// Subsequent characters can be letters or numbers
+			if (char < 'a' || char > 'z') && (char < 'A' || char > 'Z') && (char < '0' || char > '9') {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+// isValidInstanceType validates the format of an EC2 instance type.
+func isValidInstanceType(instanceType string) bool {
+	// Instance types follow the pattern: family.size (e.g., m5.large, c5.xlarge)
+	parts := strings.Split(instanceType, ".")
+	if len(parts) != 2 {
+		return false
+	}
+
+	family, size := parts[0], parts[1]
+
+	// Validate family part
+	if !isValidInstanceFamily(family) {
+		return false
+	}
+
+	// Validate size part - common sizes include: nano, micro, small, medium, large, xlarge, 2xlarge, etc.
+	validSizes := map[string]bool{
+		"nano": true, "micro": true, "small": true, "medium": true, "large": true,
+		"xlarge": true, "2xlarge": true, "3xlarge": true, "4xlarge": true, "6xlarge": true,
+		"8xlarge": true, "9xlarge": true, "10xlarge": true, "12xlarge": true, "16xlarge": true,
+		"18xlarge": true, "24xlarge": true, "32xlarge": true, "48xlarge": true, "56xlarge": true,
+		"112xlarge": true, "224xlarge": true, "metal": true,
+	}
+
+	return validSizes[size]
+}
```
good point, no need to reproduce that logic.

rvanderp3 replied:
it looks like the ec2 client isn't typically used in the context of a webhook.  perhaps we forego the webhook logic and let this fail during reconciliation.  what do you think?

enxebre commented on pkg/cloud/services/ec2/dedicatedhosts.go:
```diff
@@ -0,0 +1,260 @@
+}
+
+// ListDedicatedHosts lists dedicated hosts based on filters.
+func (s *Service) ListDedicatedHosts(ctx context.Context, filters map[string]string) ([]*infrav1.DedicatedHostInfo, error) {
```
where is this used?

rvanderp3 replied:
it was a vestige from initial work on this, it is not used and has been removed.

enxebre commented on pkg/cloud/services/ec2/dedicatedhosts.go:
```diff
@@ -0,0 +1,260 @@
+		return "", errors.Wrap(err, "failed to allocate dedicated host")
+	}
+
+	if len(output.HostIds) == 0 {
```
does aws give guarantees to always return an id immediately in the output response to the creation?
Otherwise that scenario would need to be handled with an async mechanism to fetch on going allocations

enxebre commented on pkg/cloud/services/ec2/dedicatedhosts.go:
```diff
@@ -0,0 +1,260 @@
+		return "", errors.New("no dedicated host ID returned from allocation")
+	}
+
+	hostID := output.HostIds[0]
```
why [0] what happens if quantity > 1

enxebre commented on api/v1beta2/awsmachine_types.go:
```diff
@@ -255,6 +260,40 @@ type AWSMachineSpec struct {
+type DynamicHostAllocationSpec struct {
+	// InstanceFamily specifies the EC2 instance family (e.g., "m5", "c5", "r5").
+	// +kubebuilder:validation:Required
+	InstanceFamily string `json:"instanceFamily"`
```
can we have regex cel/webhook validation to prevent invalid input for InstanceFamily / InstanceType.

Can we also compute the upper level spec.InstanceType in that validation, so any invalid input fails earlier

rvanderp3 replied:
the host family is now inferred from the machine spec

rvanderp3 commented on pkg/cloud/services/ec2/instances.go:
```diff
@@ -1276,6 +1289,79 @@ func getInstanceMetadataOptionsRequest(metadataOptions *infrav1.InstanceMetadata
+		s.scope.Debug("Using existing allocated host", "hostID", hostID)
+
+		// Validate the host is still compatible and available
+		err := s.ValidateHostCompatibility(ctx, hostID, scope.AWSMachine.Spec.InstanceType)
```
yep, that makes sense

rvanderp3 commented on pkg/cloud/services/ec2/instances.go:
```diff
@@ -1276,6 +1289,79 @@ func getInstanceMetadataOptionsRequest(metadataOptions *infrav1.InstanceMetadata
+	}
+
+	// Fallback to other AZ sources if needed
+	if availabilityZone == nil {
```
the az will not be inferred directly from the machine subnet. there shouldnt be an opportunity for a mismatch now

rvanderp3 commented on api/v1beta2/awsmachine_types.go:
```diff
@@ -246,6 +246,11 @@ type AWSMachineSpec struct {
 	// +kubebuilder:validation:Enum:=default;host
 	HostAffinity *string `json:"hostAffinity,omitempty"`
 
+	// DynamicHostAllocation enables automatic allocation of dedicated hosts.
```
for sure, will add that

rvanderp3 commented on api/v1beta2/awsmachine_types.go:
```diff
@@ -255,6 +260,40 @@ type AWSMachineSpec struct {
+
+	// Quantity specifies the number of dedicated hosts to allocate.
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=10
```
this was removed altogether

rvanderp3 commented:
/hold

investigating test failures

rvanderp3 commented:
/unhold

rvanderp3 commented on pkg/cloud/services/ec2/dedicatedhosts.go:
```diff
@@ -0,0 +1,199 @@
+	s.scope.Debug("Allocating single dedicated host", "instanceType", instanceType, "availabilityZone", availabilityZone)
+	instanceFamily := extractInstanceFamily(instanceType)
+	input := &ec2.AllocateHostsInput{
+		InstanceFamily:   aws.String(instanceFamily),
```
Good point, this may be a misunderstanding on my part.  I thought the dedicated host was associated with a family and instances are deployed on it.  Perhaps if you pass an instance type, it derives the family from that.  I'll need to investigate.

rvanderp3 commented on pkg/cloud/services/ec2/dedicatedhosts.go:
```diff
@@ -0,0 +1,199 @@
+		HostIds: []string{hostID},
+	}
+
+	_, err := s.EC2Client.ReleaseHosts(ctx, input)
+	if err != nil {
+		record.Warnf(s.scope.InfraCluster(), "FailedReleaseDedicatedHost", "Failed to release dedicated host %s: %v", hostID, err)
+		return errors.Wrap(err, "failed to release dedicated host")
+	}
```
by the time we release the dedicated host, all instances on the host should be gone.  Generally, it's a quick process.  regardless, I'm adding the retryer.

punkwalker replied:
Let's use either CAPA retryer or AWS SDK retryer whenever we need retries instead of implementing our own like this one `shouldRetryHostRelease`

punkwalker commented on pkg/cloud/services/ec2/dedicatedhosts.go:
```diff
@@ -0,0 +1,199 @@
+		HostIds: []string{hostID},
+	}
+
+	_, err := s.EC2Client.ReleaseHosts(ctx, input)
```
Based on the `ReleaseHostsOutput`, looks like the call may succeed and the host release will still be unsuccessful. So it's better to rely on output for release failures and API errors can be inferred from err returned by release call.

rvanderp3 commented on api/v1beta2/awsmachine_types.go:
```diff
@@ -255,6 +276,21 @@ type AWSMachineSpec struct {
+// This specification always allocates exactly one dedicated host per machine.
+type DynamicHostAllocationSpec struct {
+	// AutoRelease determines whether to automatically release the dedicated host
+	// when the machine is deleted.
+	// +kubebuilder:default=on-machine-deletion
+	// +kubebuilder:validation:Enum:=on-machine-deletion;manual
+	// +optional
+	Release DynamicHostReleaseStrategy `json:"release,omitempty"`
```
fair point, we can always add that later if we need to

rvanderp3 commented on api/v1beta2/awsmachine_types.go:
```diff
@@ -255,6 +276,21 @@ type AWSMachineSpec struct {
+	// +optional
+	Release DynamicHostReleaseStrategy `json:"release,omitempty"`
+
+	// Tags to apply to the allocated dedicated host.
+	// +optional
+	Tags map[string]string `json:"tags,omitempty"`
+}
```
we could for sure.  will make that change

rvanderp3 commented:
/hold 

i need to look a bit closer, it does appear one of the new unit tests is failing

rvanderp3 commented:
/unhold

mtulio reviewed (commented):
Hey @rvanderp3 , thanks for addressing the feedback. Couple of nits, and one question of failed condition.

rvanderp3 commented on controllers/awsmachine_controller.go:
```diff
@@ -360,6 +360,35 @@ func (r *AWSMachineReconciler) reconcileDelete(ctx context.Context, machineScope
+			if err := ec2Service.ReleaseDedicatedHost(ctx, hostID); err != nil {
+				// Host release failed, set up retry logic
+				machineScope.Error(err, "failed to release dedicated host", "hostID", hostID)
+				r.Recorder.Eventf(machineScope.AWSMachine, corev1.EventTypeWarning, "FailedReleaseHost", "Failed to release dedicated host %s: %v", hostID, err)
```
yes, it is, im thinking we'll keep this condition

rvanderp3 commented on controllers/awsmachine_controller.go:
```diff
@@ -360,6 +360,35 @@ func (r *AWSMachineReconciler) reconcileDelete(ctx context.Context, machineScope
+			} else {
+				// Host release succeeded
+				machineScope.Info("Successfully released dedicated host", "hostID", hostID)
+				r.Recorder.Eventf(machineScope.AWSMachine, corev1.EventTypeNormal, "SuccessfulReleaseHost", "Released dedicated host %s", hostID)
```
yep, we dont need this anymore

rvanderp3 commented on api/v1beta2/conditions_consts.go:
```diff
@@ -191,4 +195,13 @@ const (
+	DedicatedHostReleaseFailedReason = "DedicatedHostReleaseFailed"
+
+	// DedicatedHostReleaseRetryingReason used when the dedicated host release is being retried.
+	DedicatedHostReleaseRetryingReason = "DedicatedHostReleaseRetrying"
```
same, we dont need this anymore

rvanderp3 commented on pkg/cloud/services/ec2/instances.go:
```diff
@@ -1279,6 +1295,80 @@ func getInstanceMetadataOptionsRequest(metadataOptions *infrav1.InstanceMetadata
+		}
+
+		// Get the full subnet object to extract availability zone
+		subnet, err := s.getSubnet(&subnetID)
```
yep, agreed.  fixed.

mtulio reviewed (commented):
Thanks for addressing the comments, @rvanderp3 ! This looks good to me! 👍🏽 

/lgtm

rvanderp3 commented:
> > Thanks for addressing the comments, @rvanderp3 ! This looks good to me! 👍🏽
> > /lgtm
> 
> @rvanderp3 nit: you may need to squash the commits :)

good point.  done!

mtulio commented:
/lgtm

rvanderp3 commented:
/test pull-cluster-api-provider-aws-e2e-blocking

rvanderp3 commented:
/assign @punkwalker

nrb commented on api/v1beta2/awsmachine_types.go:
```diff
@@ -438,6 +461,21 @@ type AWSMachineStatus struct {
+
+// DedicatedHostStatus defines the observed state of a dynamically allocated dedicated host
+// associated with an AWSMachine. This struct is used to track the ID of the dedicated host
+// and any failure messages encountered during host release operations.
```
This part should be omitted if we're not including a field for the failure messages right now.
